Automattic/woocommerce-payments
# Add PHP compatibility checks to PRs

Fixes #3326

#### Changes proposed in this Pull Request

Add a new workflow, which runs the `PHPCompatibility` rules on both local files, and production dependencies.

We already have the `woocommerce/woocommerce-sniffs` dependency, which includes everything necessary to check for PHP compatibility, but until now it was only used for WCPay-specific code.

This PR includes 3 changes:

1. Create a new `phpcs-compat.xml.dist` file, based on the existing `phpcs.xml.dist`. The new file removes all other rules, except for `PHPCompatibility`, which we need.
2. Create a new `phpcs-compat.sh` bash script, which:
    - Does a standard Composer install, which will allow access to the `phpcs` command.
    - Performs an additional Composer install, but into a separate `vendor-dist` directory, and with the `--no-dev` flag, in order to only include production dependencies. The new directory is not ignored in the configuration, so it will be cheked.
    - Actually runs `phpcs` with the new configuration.
3. An additional workflow, which uses a minimal configuration in order to run the new script. It runs on PHP 7.4, but this does not matter, as the code sniffer will catch issues based on the `minimum_supported_wp_version` and `testVersion` values. Those are set to `7.0`, as the plugin requires this version.

Now we'll be sure what we're not introducing PHP incompatibility through WCPay, and its dependencies.

**Note:** The workflow currently fails, because there are a couple of errors. If we agree to this approach, we should ask for them to be fixed before merging this PR.

#### Testing instructions

Check how the workflow currently runs, and its output.

Additionally, you can do a local test:

1. Comment the two `rm -rf vendor-dist` lines in `bin/phpcs-compat.sh`.
2. Run `./bin/phpcs-compat.sh`. This will output some errors, ignore them for now.
3. Navigate to any PHP file within the new `vendor-dist` and introduce a PHP compatibility error.
4. Run `./bin/phpcs-compat.sh` again. There should be an error, related to the change.

Here's how this worked out for me when I added a comma at the end of a function's parameter list:


<img width="1848" alt="Screenshot 2021-11-09 at 16 17 26" src="https://user-images.githubusercontent.com/5311119/140940541-5eccfc02-8153-4c3c-a6e4-0dfc7fe383e0.png">

-------------------

- [x] No changelog entry required
- [x] Covered with tests (it is basically tests)

**Before merge**

- [ ] Make sure that all existing PHP errors/warnings are either ignored or fixed.
